File: matiasmasso/Mat.NET/DTO/DTOCustomerRanking.cs
using MatHelperStd;
using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOCustomerRanking
    {
        public DTOUser User { get; set; }
        public DTOProduct Product { get; set; }
        public DTOArea Area { get; set; }
        public DateTime FchFrom { get; set; }
        public DateTime FchTo { get; set; }

        public List<DTOCustomerRankingItem> items { get; set; }
        public decimal Sum { get; set; }

        public DTOCustomerRanking() : base()
        {
            items = new List<DTOCustomerRankingItem>();
        }

        public static DTOCustomerRanking Factory(DTOUser oUser, DTOProduct oProduct = null/* TODO Change to default(_) if this is not a reference type */, DTOArea oArea = null/* TODO Change to default(_) if this is not a reference type */, DateTime DtFchFrom = default(DateTime), DateTime DtFchTo = default(DateTime))
        {
            DTOCustomerRanking retval = new DTOCustomerRanking();
            {
                var withBlock = retval;
                withBlock.User = oUser;
                withBlock.Area = oArea;
                withBlock.Product = oProduct;
                withBlock.FchFrom = DtFchFrom == default(DateTime) ? DTO.GlobalVariables.Today().AddYears(-1) : DtFchFrom;
                withBlock.FchTo = DtFchTo == default(DateTime) ? DTO.GlobalVariables.Today() : DtFchTo;
            }
            return retval;
        }

        public void AddItem(DTOCustomer oCustomer, decimal DcEur)
        {
            DTOCustomerRankingItem item = new DTOCustomerRankingItem(oCustomer, DcEur);
            items.Add(item);
            Sum = items.Sum(x => x.Eur);
        }

        public decimal Quota(decimal DcEur)
        {
            decimal retval = 0;
            if (Sum != 0)
                retval = DcEur / Sum;
            return retval;
        }

        public decimal Accumulated(ref decimal Arrastre, decimal DcEur)
        {
            Arrastre = Arrastre + DcEur / Sum;
            decimal retval = Arrastre;
            return retval;
        }

        public string Rank(DTOCustomerRankingItem item)
        {
            int i = items.IndexOf(item) + 1;
            string retval = TextHelper.VbFormat(i, "0000");
            return retval;
        }

        public DTOCsv Csv()
        {
            string sFilename = string.Format("M+O Customer ranking {0:yyyy.MM.dd}-{1:yyyy.MM.dd}.csv", FchFrom, FchTo);
            DTOCsv retval = new DTOCsv(sFilename);
            foreach (DTOCustomerRankingItem item in items)
            {
                var oRow = retval.AddRow();
                oRow.AddCell(item.Customer.FullNom);
                oRow.AddCell(item.Eur.ToString());
            }
            return retval;
        }
    }

    public class DTOCustomerRankingItem
    {
        public DTOCustomer Customer { get; set; }
        public decimal Eur { get; set; }

        public DTOCustomerRankingItem(DTOCustomer oCustomer, decimal DcEur) : base()
        {
            Customer = oCustomer;
            Eur = DcEur;
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOCustomerRanking.vb
Public Class DTOCustomerRanking
    Property User As DTOUser
    Property Product As DTOProduct
    Property Area As DTOArea
    Property FchFrom As Date
    Property FchTo As Date

    Property items As List(Of DTOCustomerRankingItem)
    Property Sum As Decimal

    Public Sub New()
        MyBase.New()
        _items = New List(Of DTOCustomerRankingItem)
    End Sub

    Shared Function Factory(oUser As DTOUser, Optional oProduct As DTOProduct = Nothing, Optional oArea As DTOArea = Nothing, Optional DtFchFrom As Date = Nothing, Optional DtFchTo As Date = Nothing) As DTOCustomerRanking
        Dim retval As New DTOCustomerRanking
        With retval
            .User = oUser
            .Area = oArea
            .Product = oProduct
            .FchFrom = If(DtFchFrom = Nothing, DateTime.Today.AddYears(-1), DtFchFrom)
            .FchTo = If(DtFchTo = Nothing, DateTime.Today, DtFchTo)
        End With
        Return retval
    End Function

    Public Sub AddItem(oCustomer As DTOCustomer, DcEur As Decimal)
        Dim item As New DTOCustomerRankingItem(oCustomer, DcEur)
        _items.Add(item)
        _Sum = _items.Sum(Function(x) x.Eur)
    End Sub

    Public Function Quota(DcEur As Decimal) As Decimal
        Dim retval As Decimal
        If _Sum <> 0 Then
            retval = DcEur / _Sum
        End If
        Return retval
    End Function

    Public Function Accumulated(ByRef Arrastre As Decimal, DcEur As Decimal) As Decimal
        Arrastre = Arrastre + DcEur / _Sum
        Dim retval As Decimal = Arrastre
        Return retval
    End Function

    Public Function Rank(item As DTOCustomerRankingItem) As String
        Dim i As Integer = _items.IndexOf(item) + 1
        Dim retval As String = TextHelper.VbFormat(i, "0000")
        Return retval
    End Function

    Public Function Csv() As DTOCsv
        Dim sFilename As String = String.Format("M+O Customer ranking {0:yyyy.MM.dd}-{1:yyyy.MM.dd}.csv", _FchFrom, _FchTo)
        Dim retval As New DTOCsv(sFilename)
        For Each item As DTOCustomerRankingItem In _items
            Dim oRow = retval.AddRow()
            oRow.AddCell(item.Customer.FullNom)
            oRow.AddCell(item.Eur)
        Next
        Return retval
    End Function
End Class

Public Class DTOCustomerRankingItem
    Property Customer As DTOCustomer
    Property Eur As Decimal

    Public Sub New(oCustomer As DTOCustomer, DcEur As Decimal)
        MyBase.New
        _Customer = oCustomer
        _Eur = DcEur
    End Sub
End Class
